File: 07-week-6--foundational-data-structures/04-bonus--modify-linked-list-to-track-tail-and-size/solutions/linked_list.rb
class LinkedList
  attr_accessor :head
  attr_reader :size, :tail

  def initialize(head = nil)
    @head = head
    @tail = head
    @size = 0

    set_size_and_tail
  end

  def set_size_and_tail
    @size = 0

    iterate do |node|
      @tail = node
      update_size('+')
    end
  end

  def update_size(operation)
    operation === '+' ? @size += 1 : @size -= 1
  end

  def iterate
    count = 0
    temp = @head

    until temp.nil?
      yield(temp, count)      
      temp = temp.next_node
      count += 1
    end

    @head
  end

  def print
    iterate { |node| puts node.value }
  end

  def find(target)
    iterate do |node|
      return node if node.value == target
    end

    nil
  end

  def add_first(node)
    if @head.nil?
      @tail = node
    end

    node.next_node = @head
    @head = node

    update_size('+')
  end

  def add_last(node)
    if @head.nil?
      add_first(node)
      return
    end

    @tail.next_node = node
    @tail = node
    update_size('+')

    # if @head.nil?
    #   @head = node
    # end

    # iterate do |curr_node|
    #   if curr_node.next_node.nil?
    #     curr_node.next_node = node
    #     return
    #   end
    # end
  end

  def remove_first
    old_head = @head

    unless @head.nil?
      @head = @head.next_node
      update_size('-')
    end
    
    if @head.nil?
      @tail = nil
    end

    old_head
  end

  def remove_last
    return remove_first if @head.nil? || @head.next_node.nil?

    update_size('-')

    iterate do |node|
      if node.next_node.next_node.nil?
        old_tail = node.next_node
        node.next_node = nil
        @tail = node
        return old_tail
      end
    end
  end

  def replace(idx, node)
    if idx.zero?
      remove_first
      add_first(node)
      return node
    end

    iterate do |curr_node, count|
      if count == idx - 1
        node.next_node = curr_node.next_node.next_node
        curr_node.next_node = node

        @tail = node if node.next_node.nil?

        return node
      end
    end
  end

  def insert(idx, node)
    if idx.zero?
      add_first(node)
      return
    end

    update_size('+')

    iterate do |curr_node, count|
      if count == idx - 1
        old_next = curr_node.next_node
        curr_node.next_node = node
        node.next_node = old_next

        @tail = node if node.next_node.nil?

        return
      end
    end
  end

  def remove(idx)
    return remove_first if idx.zero?

    update_size('-')

    iterate do |node, count|
      if count == idx - 1
        old_node = node.next_node
        node.next_node = node.next_node.next_node

        @tail = node if node.next_node.nil?

        return old_node
      end
    end
  end

  def clear
    @head = nil
    @tail = nil
    @size = 0
  end
end

class Node
  attr_accessor :value, :next_node

  def initialize(value = nil, next_node = nil)
    @value = value
    @next_node = next_node
  end
end

if __FILE__ == $PROGRAM_NAME
  head = Node.new('one', Node.new('two', Node.new('three', Node.new('four'))))
  list = LinkedList.new(head)
  empty_list = LinkedList.new
  one_item_list = LinkedList.new(Node.new('just one'))
  
  puts("CHECK FOR CORRECT TAIL ON INIT");
  puts("Should be 'four': #{list.tail.value}");
  puts("Should be 'null': #{empty_list.tail}");
  puts("Should be 'just one': #{one_item_list.tail.value}");
  puts("---------------------------------------------------");

  puts("CHECK FOR CORRECT TAIL ON ADD FIRST");
  list.add_first(Node.new('beans'))
  empty_list.add_first(Node.new('beans'))
  one_item_list.add_first(Node.new('beans'))
  puts("Should be 'four': #{list.tail.value}")
  puts("Should be 'beans': #{empty_list.tail.value}")
  puts("Should be 'just one': #{one_item_list.tail.value}")
  [list, empty_list, one_item_list].each { |l| l.remove_first }
  puts("---------------------------------------------------")

  puts("CHECK FOR CORRECT TAIL ON ADD LAST")
  list.add_last(Node.new('beans'))
  empty_list.add_last(Node.new('beans'))
  one_item_list.add_last(Node.new('beans'))
  puts("Should be 'beans': #{list.tail.value}")
  puts("Should be 'beans': #{empty_list.tail.value}")
  puts("Should be 'beans': #{one_item_list.tail.value}")
  [list, empty_list, one_item_list].each do |l|
    l.remove_last
    l.set_size_and_tail
  end
  puts("---------------------------------------------------")

  puts("CHECK FOR CORRECT TAIL ON REMOVE FIRST")
  removed = list.remove_first
  removed_again = one_item_list.remove_first
  empty_list.remove_first
  puts("Should be 'four': #{list.tail.value}")
  puts("Should be 'null': #{empty_list.tail}")
  puts("Should be 'null': #{one_item_list.tail}")
  list.add_first(removed)
  one_item_list.add_first(removed_again)
  puts("---------------------------------------------------")

  puts("CHECK FOR CORRECT TAIL ON REMOVE LAST")
  removed = list.remove_last
  removed_again = one_item_list.remove_last
  empty_list.remove_last
  puts("Should be 'three': #{list.tail.value}")
  puts("Should be 'null': #{empty_list.tail}")
  puts("Should be 'null': #{one_item_list.tail}")
  list.add_last(removed)
  one_item_list.add_last(removed_again)
  puts("---------------------------------------------------")

  puts("CHECK FOR CORRECT TAIL ON REPLACE")
  node_two = Node.new('two')
  node_one = Node.new('one', node_two)
  replace_list = LinkedList.new(node_one)

  replace_list.replace(0, Node.new('hi'))
  puts("Should be two: #{replace_list.tail.value}")
  replace_list.replace(1, Node.new('bye'))
  puts("Should be bye: #{replace_list.tail.value}")
  puts("---------------------------------------------------")

  puts("CHECK FOR CORRECT TAIL ON INSERT")
  node_two = Node.new('two')
  node_one = Node.new('one', node_two)
  insert_list = LinkedList.new(node_one)

  insert_list.insert(0, Node.new('hi'))
  # hi -> one -> two
  puts("Should be two: #{insert_list.tail.value}")
  insert_list.insert(1, Node.new('bye'))
  # hi -> bye -> one -> two
  puts("Should be two: #{insert_list.tail.value}")
  insert_list.insert(4, Node.new('last'))
  # hi -> bye -> one -> two -> last
  puts("Should be last: #{insert_list.tail.value}")
  puts("---------------------------------------------------")

  puts("CHECK FOR CORRECT TAIL ON REMOVE")
  # one -> two -> three -> four
  list.remove(1)
  # one -> three -> four
  puts("Should be four: #{list.tail.value}")
  list.remove(2)
  # one -> three
  puts("Should be three: #{list.tail.value}")
  list.insert(1, Node.new('two'))
  list.add_last(Node.new('four'))
  puts("---------------------------------------------------")

  puts("CHECK FOR CORRECT TAIL ON CLEAR")
  list.clear
  puts("Should be null: #{list.tail}")
  head = Node.new('one', Node.new('two', Node.new('three', Node.new('four'))))
  list = LinkedList.new(head)
  puts("---------------------------------------------------")

  puts("CHECKING SIZE")
  puts("New empty list should be 0: #{LinkedList.new.size}")
  puts("New list with one Node should be 1: #{LinkedList.new(Node.new('hi')).size}")
  puts("New list with 3 nodes should be 3: #{LinkedList.new(Node.new('hi', Node.new('bye', Node.new('what')))).size}")
  list.add_first(Node.new('s'))
  puts("On addFirst should be 5: #{list.size}")
  list.add_last(Node.new('b'))
  puts("On addLast should be 6: #{list.size}")
  list.remove_first
  puts("On removeFirst should be 5: #{list.size}")
  list.remove_last
  puts("On removeLast should be 4: #{list.size}")
  list.insert(1, Node.new('stuff'))
  puts("On insert should be 5: #{list.size}")
  list.remove(1)
  puts("On remove should be 4: #{list.size}")
  list.clear
  puts("On clear should be 0: #{list.size}")
end

File: 07-week-6--foundational-data-structures/04-bonus--modify-linked-list-to-track-tail-and-size/solutions/linked_list.js
class LinkedList {
  constructor(head = null) {
    this.head = head;
    this.tail = null;
    this.size = 0;
    // worst case O(n) for setting tail on initializing new list
    this.setTailAndSize();
  }

  // helper method for setting tail when list goes from empty to not empty
  // upon initialization
  // will be O(n) because user could add Node connected to other Nodes
  setTailAndSize() {
    this.size = 0;

    this.iterate((node) => {
      this.tail = node;
      this.updateSize('+');
    });
  }

  // helper method for updating size
  updateSize(operation) {
    operation === '+' ? ++this.size : --this.size;
  }

  iterate(callback) {
    let count = 0;
    let temp = this.head;

    while (temp !== null) {
      const result = callback(temp, count);

      if (result === true) {
        return temp;
      }

      ++count;
      temp = temp.next;
    }

    return this.head;
  }

  // print each node's value on its own line
  // use your iterate method to be DRY! Don't get caught in the code rain, brrr.
  print() {
    this.iterate(node => console.log(node.value));
  }

  // find the node with the target value and return it
  // if not found return null, use your iterate method to be DRY!
  find(target) {
    let result = null;

    this.iterate(node => {
      if (node.value === target) {
        result = node;

        return true;
      }
    });

    return result;
  }

  // add the node to the start of the list, no nodes should be removed
  addFirst(node) {
    // O(1) operation: the tail will only be set if the list is empty
    if (this.head === null) {
      this.tail = node;
    }

    node.next = this.head;
    this.head = node;

    // increase size: O(1)
    this.updateSize('+');
  }

  // add node to end of list, no nodes should be removed
  // you may wish to use the iterate method
  addLast(node) {
    if (this.head === null) {
      this.addFirst(node);
      return;
    }

    this.tail.next = node;
    // O(1) operation to set tail
    this.tail = node;
    // increase size: O(1)
    this.updateSize('+');

    // if (this.head === null) {
    //   this.head = node;
    // }

    // this.iterate(currNode => {
    //   if (currNode.next === null) {
    //     currNode.next = node;
    //     return true;
    //   }
    // });
  }

  // remove the first Node in the list and update head
  // and return the removed node
  removeFirst() {
    const oldHead = this.head;

    if (this.head !== null) {
      this.head = this.head.next;
      this.updateSize('-');
    }

    // if list goes from 1 Node to empty, update tail O(1)
    if (this.head === null) {
      this.tail = null;
    }

    return oldHead;
  }

  // remove the tail node, iterate may be helpful
  // return the node you just removed
  removeLast() {
    if (this.head === null || this.head.next === null) {
      return this.removeFirst();
    }

    let oldTail = null;

    this.iterate(node => {
      if (node.next.next === null) {
        oldTail = node.next;
        node.next = null;
        return true;
      }
    });

    // Could also set the tail inside iterate above
    // O(n) operation
    this.setTailAndSize();

    return oldTail;
  }

  // replace the node at the given index with the given node
  replace(idx, node) {
    if (idx === 0) {
      this.removeFirst();
      this.addFirst(node);
      return node;
    }

    this.iterate((currNode, count) => {
      if (count === idx - 1) {
        node.next = currNode.next.next;
        currNode.next = node;

        // If tail is being replaced O(1)
        if (node.next === null) {
          this.tail = node;
        }

        return true;
      }
    });

    return node;
  }

  // insert the node at the given index
  // no existing nodes should be removed or replaced
  insert(idx, node) {
    if (idx === 0) {
      this.addFirst(node);
      return;
    }

    this.iterate((currNode, count) => {
      if (count === idx - 1) {
        const oldNext = currNode.next;
        currNode.next = node;
        node.next = oldNext;

        // if new node is inserted at the very end, O(1)
        if (oldNext === null) {
          this.tail = node;
        }

        return true;
      }
    });

    this.updateSize('+');
  }

  // remove the node at the given index, and return it
  remove(idx) {
    if (idx === 0) {
      return this.removeFirst();
    }

    let oldNode = null;

    this.iterate((node, count) => {
      if (count === idx - 1) {
        oldNode = node.next;
        node.next = node.next.next;
        // if tail was removed, update tail
        if (node.next === null) {
          this.tail = node;
        }

        return true;
      }
    }); 

    this.updateSize('-');

    return oldNode;
  }

  clear() {
    this.head = null;
    // null tail for empty list
    this.tail = null;
    this.size = 0;
  }
}

class Node {
  constructor(value = null, next = null) {
    this.value = value;
    this.next = next;
  }
}

if (require.main === module) {
  let head = new Node('one', new Node('two', new Node('three', new Node('four'))));
  let list = new LinkedList(head);
  let emptyList = new LinkedList();
  let oneItemList = new LinkedList(new Node('just one'));

  console.log("CHECK FOR CORRECT TAIL ON INIT");
  console.log(`Should be 'four': ${list.tail.value}`);
  console.log(`Should be 'null': ${emptyList.tail}`);
  console.log(`Should be 'just one': ${oneItemList.tail.value}`);
  console.log("---------------------------------------------------");

  console.log("CHECK FOR CORRECT TAIL ON ADD FIRST");
  list.addFirst(new Node('beans'));
  emptyList.addFirst(new Node('beans'));
  oneItemList.addFirst(new Node('beans'));
  console.log(`Should be 'four': ${list.tail.value}`);
  console.log(`Should be 'beans': ${emptyList.tail.value}`);
  console.log(`Should be 'just one': ${oneItemList.tail.value}`);
  [list, emptyList, oneItemList].forEach(l => l.removeFirst());
  console.log("---------------------------------------------------");

  console.log("CHECK FOR CORRECT TAIL ON ADD LAST");
  list.addLast(new Node('beans'));
  emptyList.addLast(new Node('beans'));
  oneItemList.addLast(new Node('beans'));
  console.log(`Should be 'beans': ${list.tail.value}`);
  console.log(`Should be 'beans': ${emptyList.tail.value}`);
  console.log(`Should be 'beans': ${oneItemList.tail.value}`);
  [list, emptyList, oneItemList].forEach(l => {
    l.removeLast();
    l.setTailAndSize();
  });
  console.log("---------------------------------------------------");

  console.log("CHECK FOR CORRECT TAIL ON REMOVE FIRST");
  let removed = list.removeFirst();
  let removedAgain = oneItemList.removeFirst();
  emptyList.removeFirst();
  console.log(`Should be 'four': ${list.tail.value}`);
  console.log(`Should be 'null': ${emptyList.tail}`);
  console.log(`Should be 'null': ${oneItemList.tail}`);
  list.addFirst(removed);
  oneItemList.addFirst(removedAgain);
  console.log("---------------------------------------------------");

  console.log("CHECK FOR CORRECT TAIL ON REMOVE LAST");
  removed = list.removeLast();
  removedAgain = oneItemList.removeLast();
  emptyList.removeLast();
  console.log(`Should be 'three': ${list.tail.value}`);
  console.log(`Should be 'null': ${emptyList.tail}`);
  console.log(`Should be 'null': ${oneItemList.tail}`);
  list.addLast(removed);
  oneItemList.addLast(removedAgain);
  console.log("---------------------------------------------------");

  console.log("CHECK FOR CORRECT TAIL ON REPLACE");
  let nodeTwo = new Node('two');
  let nodeOne = new Node('one', nodeTwo);
  let replaceList = new LinkedList(nodeOne);

  replaceList.replace(0, new Node('hi'));
  console.log(`Should be two: ${replaceList.tail.value}`);
  replaceList.replace(1, new Node('bye'));
  console.log(`Should be bye: ${replaceList.tail.value}`);
  console.log("---------------------------------------------------");

  console.log("CHECK FOR CORRECT TAIL ON INSERT");
  nodeTwo = new Node('two');
  nodeOne = new Node('one', nodeTwo);
  insertList = new LinkedList(nodeOne);

  insertList.insert(0, new Node('hi'));
  // hi -> one -> two
  console.log(`Should be two: ${insertList.tail.value}`);
  insertList.insert(1, new Node('bye'));
  // hi -> bye -> one -> two
  console.log(`Should be two: ${insertList.tail.value}`);
  insertList.insert(4, new Node('last'));
  // hi -> bye -> one -> two -> last
  console.log(`Should be last: ${insertList.tail.value}`);
  console.log("---------------------------------------------------");

  console.log("CHECK FOR CORRECT TAIL ON REMOVE");
  // one -> two -> three -> four
  list.remove(1);
  // one -> three -> four
  console.log(`Should be four: ${list.tail.value}`);
  list.remove(2);
  // one -> three
  console.log(`Should be three: ${list.tail.value}`);
  list.insert(1, new Node('two'));
  list.addLast(new Node('four'));
  console.log("---------------------------------------------------");

  console.log("CHECK FOR CORRECT TAIL ON CLEAR");
  list.clear();
  console.log(`Should be null: ${list.tail}`);
  head = new Node('one', new Node('two', new Node('three', new Node('four'))));
  list = new LinkedList(head);
  console.log("---------------------------------------------------");

  console.log("CHECKING SIZE");
  console.log(`New empty list should be 0: ${new LinkedList().size}`);
  console.log(`New list with one Node should be 1: ${new LinkedList(new Node('hi')).size}`);
  console.log(`New list with 3 nodes should be 3: ${new LinkedList(new Node('hi', new Node('bye', new Node('what')))).size}`);
  list.addFirst(new Node('s'));
  console.log(`On addFirst should be 5: ${list.size}`);
  list.addLast(new Node('b'));
  console.log(`On addLast should be 6: ${list.size}`);
  list.removeFirst();
  console.log(`On removeFirst should be 5: ${list.size}`);
  list.removeLast();
  console.log(`On removeLast should be 4: ${list.size}`);
  list.insert(1, new Node('stuff'));
  console.log(`On insert should be 5: ${list.size}`);
  list.remove(1);
  console.log(`On remove should be 4: ${list.size}`);
  list.clear();
  console.log(`On clear should be 0: ${list.size}`);
}

module.exports = {
  Node, LinkedList
};
